src/heatmap.ts

```diff
@@ -0,0 +1,289 @@
+                    })
+                })
+            })
+        })
+        // Create heatmap objects that show modifiers for each allied and rival unit pair
+        let alliedRivalPairs = rivalUnits.map(rivalUnit => {
+            var obj = undefined
+            // Check whether rival unit is associated with any modifier for the allied unit
+            // Save unit pair and modifier value if true
+            // Update maxBonus if modifier value is larger
+            modifiers.forEach(m => {
+                if (rivalUnit.classes.includes(m.class)) {
+                    obj = {
+                        uX: rivalUnit.baseId,
+                        uY: alliedUnit.baseId,
+                        uV: m.value
+                    }
+                    if (m.value > maxBonus) {
+                        maxBonus = m.value
+                    }
+                }
+            })
+
+            // If rival unit is not associated with any modifier for the allied unit, save the unit pair with a value of 0
+            if (obj == undefined) {
+                obj = {
+                    uX: rivalUnit.baseId,
+                    uY: alliedUnit.baseId,
+                    uV: 0
+                }
+            }
+            // Return modifier unit pair
+            return obj;
+        }).flat()
+        // Concatenate list of pairs as heatmap data
+        heatmapData = heatmapData.concat(alliedRivalPairs);
+    })
+    // return max value from relevant modifiers and the heatmap data
+    return [maxBonus, heatmapData]
+}
+
+export async function createHeatmap(data: Unit[], allyCiv: string, rivalCiv: string) {
+    // Set the dimensions and margins of the graph
+    var margin = {top: 20, right: 5, bottom: 160, left: 170},
+        width = 900 - margin.left - margin.right,
+        height = 800 - margin.top - margin.bottom;
+
+    // Append the svg object to the body of the page
+    var svg = d3.select("#unitHeatmap")
+        .append("svg")
+        .attr("width", width + margin.left + margin.right)
+        .attr("height", height + margin.top + margin.bottom)
+        .append("g")
+        .attr("transform",
+            "translate(" + margin.left + "," + margin.top + ")")
+
+    // Labels of row and columns
+    // Create array of unique units for chosen rival and allied civilizations
+    const rivalUnits = data.filter(item => item.civs.includes(rivalCiv))
+    const uniqueRivalUnits = [
+        ...new Map(rivalUnits.map((item) => [item.baseId, item])).values(),
+    ];
+    const rivalUnitTitles = uniqueRivalUnits.map(item => item.baseId);
+    rivalUnitTitles.sort()
+
+    const alliedUnits = data.filter(item => item.civs.includes(allyCiv))
+    const uniqueAlliedUnits = [
+        ...new Map(alliedUnits.map((item) => [item.baseId, item])).values(),
+    ];
+    const alliedUnitTitles = [...new Set(alliedUnits.map(item => item.baseId))];
+    alliedUnitTitles.sort()
+
+    // Build X scales and axis:
+    var x = d3.scaleBand()
```

Comment:
Not an issue, but in general if you use `let` instead of `var` you are less likely to encounter unexpected behaviour.

Reply:
Alright thanks for letting me know